Client/Frontend/Share/ShareExtensionHelper.swift

```diff
@@ -54,27 +74,69 @@ class ShareExtensionHelper: NSObject {
         // which is after the user taps the button. So a million cycles away.
         findLoginExtensionItem()
 
-        activityViewController.completionWithItemsHandler = { activityType, completed, returnedItems, activityError in
-            if !completed {
-                completionHandler(completed, activityType)
-                return
-            }
-            // Bug 1392418 - When copying a url using the share extension there are 2 urls in the pasteboard.
-            // This is a iOS 11.0 bug. Fixed in 11.2
-            if UIPasteboard.general.hasURLs, let url = UIPasteboard.general.urls?.first {
-                UIPasteboard.general.urls = [url]
-            }
+        activityViewController.completionWithItemsHandler = { [weak self] activityType, completed, returnedItems, activityError in
+            guard let self = self else { return }
             
-            if self.isPasswordManagerActivityType(activityType.map { $0.rawValue }) {
-                if let logins = returnedItems {
-                    self.fillPasswords(logins as [AnyObject])
+            if self.shareActivityType(activityType.map { $0.rawValue }) != .iBooks { //!self.isOpenInIBooksActivityType(activityType.map { $0.rawValue }) {
+                if !completed {
+                    completionHandler(completed, activityType, nil)
+                    return
+                }
+                // Bug 1392418 - When copying a url using the share extension there are 2 urls in the pasteboard.
+                // This is a iOS 11.0 bug. Fixed in 11.2
+                if UIPasteboard.general.hasURLs, let url = UIPasteboard.general.urls?.first {
+                    UIPasteboard.general.urls = [url]
+                }
+                
+                if self.shareActivityType(activityType.map { $0.rawValue }) != .password { //self.isPasswordManagerActivityType(activityType.map { $0.rawValue }) {
+                    if let logins = returnedItems {
+                        self.fillPasswords(logins as [AnyObject])
+                    }
                 }
-            }
 
-            completionHandler(completed, activityType)
+                completionHandler(completed, activityType, nil)
+            } else {
+                self.writeWebPagePDFDataToURL { url, error in
+                    completionHandler(completed, activityType, url)
+                }
+            }
         }
+        
         return activityViewController
     }
+    
+    func writeWebPagePDFDataToURL(_ completion: @escaping (URL?, Error?) -> Void) {
+        #if compiler(>=5.3)
```

Comment:
this means save pdf to iBooks works only on iOS14+?

Reply:
It shouldnt be but the only solution I managed to find out is to create the pdf and get the link and pass it to books app to open. I might have to come back to this If we re aiming for < iOS 14 solution.

Reply:
from what I've read in the corresponding ticket, it works in Firefox, can you check it and see if it works on FF pre iOS14?

Reply:
Ok I misunderstood here, I thought we want to share any website as pdf using iBooks. But actually we want to share pdf file opened in browser in iBooks.

- Sharing any website as pdf using iBooks also fails in Firefox however sharing pdf file opened in browser doesnt fail 

I will look into 2nd pdf file sharing in detail and add it to this PR.